# What is Fake News?

A type of yellow journalism, fake news encapsulates pieces of news that may be hoaxes and is generally spread through social media and other online media. This is often done to further or impose certain ideas and is often achieved with political agendas. Such news items may contain false and/or exaggerated claims, and may end up being viralized by algorithms, and users may end up in a filter bubble.



# Objective :

## Detecting Fake News with Python 
To build a model to accurately classify a piece of news as REAL or FAKE.


## About the Python Project
This advanced python project of detecting fake news deals with fake and real news. Using sklearn, we build a TfidfVectorizer on our dataset. Then, we initialize a PassiveAggressive Classifier and fit the model. In the end, the accuracy score and the confusion matrix tell us how well our model fares.

 

## The Dataset
The dataset we’ll use for this python project- we’ll call it news.csv. This dataset has a shape of 7796×4. The first column identifies the news, the second and third are the title and text, and the fourth column has labels denoting whether the news is REAL or FAKE. The dataset takes up 29.2MB of space.

In [1]:
import numpy as np                  
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score , confusion_matrix  
np.random.seed(43)

In [2]:
data=pd.read_csv(r"C:/Users/Shikhar/Desktop/Detecting_fake_news/news.csv")   #import the csv file as dataframe 

In [3]:
data.head() #to print first 5 rows of data to get the better insight of the data


Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [4]:
#Getting the X and Y variables ready 
y=data['label']   #inserting the values of lables column into  y
y.head(5)
X=data['text']    #inserting the values of text column into x 
X.head()

0    Daniel Greenfield, a Shillman Journalism Fello...
1    Google Pinterest Digg Linkedin Reddit Stumbleu...
2    U.S. Secretary of State John F. Kerry said Mon...
3    — Kaydee King (@KaydeeKing) November 9, 2016 T...
4    It's primary day in New York and front-runners...
Name: text, dtype: object

In [10]:
X_train,X_test,y_train,y_test = train_test_split( X , y ,test_size = 0.2,random_state = 7) #dividing the data into train  and test 

### Let’s initialize a TfidfVectorizer

With stop words from the English language and a maximum document frequency of 0.7 (terms with a higher document frequency will be discarded). Stop words are the most common words in a language that are to be filtered out before processing the natural language data. And a TfidfVectorizer turns a collection of raw documents into a matrix of TF-IDF features.



In [11]:
tfidf_vect=TfidfVectorizer(stop_words='english',max_df=0.7)

tfidf_train = tfidf_vect.fit_transform(X_train)
tfidf_test = tfidf_vect.transform(X_test)

In [12]:
pac = PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,y_train)

y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f"Accuracy is {round(score*100,2)} %")


Accuracy is 92.74 %


### We got an accuracy of around 92 % with this model.
Finally, let’s print out a confusion matrix to gain insight into the number of false and true negatives and positives.`

In [13]:
confusion_matrix(y_test,y_pred,labels=['FAKE','REAL'])



array([[589,  49],
       [ 43, 586]], dtype=int64)

#### So with this model, we have 589 true positives, 587 true negatives, 42 false positives, and 49 false negatives.

### Summary
Today, we learned to detect fake news with Python. We took a political dataset, implemented a TfidfVectorizer, initialized a PassiveAggressiveClassifier, and fit our model. 
We ended up obtaining an accuracy of 92.82% in magnitude.